### PR TITLE
Allow editing of lesson group fields in script editor gui

### PR DESCRIPTION
**Pull request by dmcavoy**
Adds the ability to edit the Lesson Group name, description and big questions on the LessonGroupCard in the Script Editor GUI.  The input/textareas only show if the Lesson Group is user facing. Otherwise it shows that you need to make the LessonGroup user facing and provides a button to do so.

### Not User Facing
Screen Shot 2020-09-29 at 2 39 16 PM

### User Facing
Screen Shot 2020-09-29 at 2 39 30 PM

## Links

- [jira](https://codedotorg.atlassian.net/browse/PLAT-268)

## Testing story

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/test/unit/lib/levelbuilder/script-editor/LessonGroupCardTest.jsx [hunk 1]
```diff
@@ -13,11 +13,18 @@ const defaultProps = {
...
+  setTargetLessonGroup: () => {},
+  targetLessonGroupPos: null,
+  generateLessonGroupKey: () => {},
   lessonGroup: {
     key: 'lg-key',
     display_name: 'Display Name',
     position: 1,
     user_facing: true,
+    description: 'Lesson group description',
+    big_questions: '',
```
Comment (davidsbailey) [inline apps/test/unit/lib/levelbuilder/script-editor/LessonGroupCardTest.jsx]: now seems like a good time to add a test showing that these two fields appear in the shallow render.

**Review by davidsbailey — COMMENTED** [review]
The UX seems like it might be a bit confusing regarding "user facing". this implementation mirrors what's in the database, but I think we could probably offer a simpler mental model to the editor. 

In the first screenshot, when there is only one non-user-facing lesson group, what do you think about taking away the "user facing" button on the lesson group, and replacing the "add lesson group" button on the Unit card with a button named "enable lesson groups"?

In this state, you could also omit this header row:
Screen Shot 2020-10-02 at 11 40 48 AM

usual caveat: I am not a UX designer. :-) any thoughts? any input, @tess323 ?

**Comment by dmcavoy** [comment]
> The UX seems like it might be a bit confusing regarding "user facing". this implementation mirrors what's in the database, but I think we could probably offer a simpler mental model to the editor.
> 
> In the first screenshot, when there is only one non-user-facing lesson group, what do you think about taking away the "user facing" button on the lesson group, and replacing the "add lesson group" button on the Unit card with a button named "enable lesson groups"?
> 
> In this state, you could also omit this header row:
> Screen Shot 2020-10-02 at 11 40 48 AM
> 
> usual caveat: I am not a UX designer. :-) any thoughts? any input, @tess323 ?

I talked to Tess and she agreed with your thoughts on the design. We also talked about preventing you from ever having 0 lesson groups. Instead if you are starting a new script or delete all the existing lesson groups we will have a non-user facing lesson group already provided for you. I'm going to work on these changes and I will update the PR.

**Comment by dmcavoy** [comment]
@tess323 @davidsbailey  Updated!

user-facing-lesson-group

**Comment by davidsbailey** [comment]
nice improvements! top-level comment: I would suggest hiding the up / down / delete controls in the case where lesson groups are not enabled:
Screen Shot 2020-10-05 at 2 06 18 PM

File: apps/test/unit/code-studio/scriptEditorReduxTest.js [hunk 2]
```diff
@@ -52,7 +68,25 @@ describe('scriptEditorRedux reducer tests', () => {
...
+
+  it('remove group', () => {
+    // Remove lesson group when there are 2 lessons groups
+    let nextState = reducer(initialState, removeGroup(1));
+    let lessonGroups = nextState.lessonGroups;
+    assert.equal(lessonGroups.length, 1);
+    assert.equal(lessonGroups[0].key, 'lg-key-2');
+
+    // Remove lesson group when there is only one lesson group left
+    // a non-user facing lesson group should be added
```
Comment (davidsbailey) [inline apps/test/unit/code-studio/scriptEditorReduxTest.js:82]: If I am understanding correctly then I'm a bit worried about this user flow - if i enable lesson groups, then realize I've made a mistake, it seems much more likely that I'll want to convert the one remaining lesson group back to non-user-facing, as opposed to deleting all my lessons. I don't have an exact solution in mind, but accidentally clicking "enable" seems like a bit of a land mine for curriculum writers.

File: apps/src/sites/studio/pages/scripts/edit.js [hunk 3]
```diff
@@ -7,23 +7,24 @@ import {getStore, registerReducers} from '@cdo/apps/redux';
...
+  empty_non_user_facing_group
 } from '@cdo/apps/lib/levelbuilder/script-editor/scriptEditorRedux';
 import ScriptEditor from '@cdo/apps/lib/levelbuilder/script-editor/ScriptEditor';
 import {valueOr} from '@cdo/apps/utils';
 
 export default function initPage(scriptEditorData) {
   const scriptData = scriptEditorData.script;
   const lessonLevelData = scriptEditorData.lessonLevelData;
-  const lesson_groups = (scriptData.lesson_groups || [])
+  let lesson_groups = (scriptData.lesson_groups || [])
```
Comment (davidsbailey) [inline apps/src/sites/studio/pages/scripts/edit.js]: please use camelCase for variable names, here and below. 

ideally, snake_case shouldn't make it any farther into our system than this "entry point". in cases where we have already let snake case into redux, no need to fix that right now, but in the future we may go back and try to push snake case further out of js.

File: apps/src/lib/levelbuilder/script-editor/scriptEditorRedux.js [hunk 4]
```diff
@@ -256,3 +261,13 @@ export default {
   levelKeyList,
   lessonGroups
 };
+
+export const empty_non_user_facing_group = {
```
Comment (davidsbailey) [inline apps/src/lib/levelbuilder/script-editor/scriptEditorRedux.js]: please use camel case

**Review by davidsbailey — COMMENTED** [review]
Nice progress Dani! A few more comments.

File: apps/test/unit/code-studio/scriptEditorReduxTest.js (hunk 2 above)
Comment (dmcavoy) [reply apps/test/unit/code-studio/scriptEditorReduxTest.js:82]: I'll look at how hard it is to have a disable lesson groups button if there is only one lesson group

Comment (davidsbailey) [reply apps/test/unit/code-studio/scriptEditorReduxTest.js:82]: ok, this would be fine to resolve in a future PR / discussion if that's easier.

**Review by davidsbailey — APPROVED** [review]

**Comment by dmcavoy** [comment]
@davidsbailey @tess323 How does this flow look?

user-facing-lesson-group

**Comment by davidsbailey** [comment]
❤️ 🚀 👍  awesome work Dani! This seals off all the problem cases I was worried about.

**Review by davidsbailey — APPROVED** [review]

**Pull request merged**